`<RollSelector/>` rich formatting and adding label numbers

simonwiles opened the pull request:
This PR fixes a couple of incidental issues (the first two commits), and then adds some rich styling options to the dropdown list items and the input field itself.  It does this by proving a `postMarkup()` hook that allows parent components to alter the display of the various elements *after* they've been filtered and marked up for matches.  In order to make this possible for the input box too, the `<input type="text"/>` element has been replaced with a `<span contenteditable="true"/>`.

This gives lots of flexibility, and from a purely aesthetic point of view I like the way it looks now.  However, the way I've split up the label number is probably controversial/flat-out wrong.  The numbers don't appear to conflict at all, but I assume that's just a fluke?  Anyway, I assume we'll probably need to adjust this somehow (or do it in a slightly different way altogether).  Do you have a clear steer to offer on this, or does this need to be a conversation involving the PIs?

Edit: There are also a couple where the `label` field in the catalogue has only a number, which currently results in an empty `[]` at the end of the line.  In the unlikely event that the current arrangement is acceptable, I'll fix that.

broadwell commented:
The rich formatting looks great! I do think the PIs probably will prefer the label number and company name to appear together, perhaps to the right of the other info, and still in square brackets, e.g., `[Welte-Mignon 520]`. but I'd be happy to merge this as is and then ask them for feedback.
First, though, I've noticed another corner case with the search field (also present in the previous PR): typing a bunch of spaces or other punctuation after an initial letter can quickly trigger memory exhaustion. Do any mitigation strategies come to mind? It's not something anyone would type intentionally, but it does highlight the potential that a sticky key could end up crashing the page...
![Screen Shot 2021-07-07 at 1 00 56 PM](https://user-images.githubusercontent.com/7329112/124833620-522dee00-df33-11eb-8eb0-8322beae5612.png)
![Screen Shot 2021-07-07 at 1 51 08 PM](https://user-images.githubusercontent.com/7329112/124833629-54904800-df33-11eb-8cd9-4c378cfc889b.png)
![Screen Shot 2021-07-07 at 12 46 16 PM](https://user-images.githubusercontent.com/7329112/124833639-578b3880-df33-11eb-8b83-3d95bcbaf8e0.png)

simonwiles commented:
Yep, thanks.  I can reproduce but only when I add another (non-whitespace, non-punctuation) character *after* the long string of spaces/punctuation (and understand this and have a fix for this ready) -- is that the same for you?

broadwell commented:
> Yep, thanks. I can reproduce but only when I add another (non-whitespace, non-punctuation) character _after_ the long string of spaces/punctuation (and understand this and have a fix for this ready) -- is that the same for you?

OK yeah, I forgot to include that detail: the warning only appears when I try to type another character after what you see in the screenshots (but the character doesn't appear, due to the resource exhaustion).

simonwiles commented:
Okay, great, thanks -- fix(es) committed.

The algorithm involves `searchString.split(" ")` and 
```
"a              b".split(" ") === ["a", "", "", "", "", "", "", "", "", "", "", "", "", "", "b"]
```
(and punctuation is replaced with white space, hence the same behaviour with `a...................b`).

The first extra commit collapses consecutive whitespace (after the punctuation substitution), and the second enforces an arbitrary limit of 8 search terms, with any excess being silently discarded.

broadwell commented:
Cool -- I was thinking that an arbitrary search term limit might be called for here.
Regarding the formatting: we'll probably want to add some logic to handle empty values, see e.g., this 88-note roll with an issue number but no company name:
![Screen Shot 2021-07-07 at 3 46 59 PM](https://user-images.githubusercontent.com/7329112/124838288-5f4edb00-df3b-11eb-8294-b015321be4f0.png)
Probably this can wait until we get feedback about the desired format of the dropdown items.

simonwiles commented:
Yep, agreed -- see the note in my original "post".  I think it probably makes sense to merge this first, though, and deal with changes later, if that's cool with you?  (want to get the faceting in, too)

broadwell commented:
OK right, I see the relevant section in your comment! Let's merge this now and deal with the empty values when we have a better handle on the desired format.

simonwiles commented:
Thanks -- like you I have a feeling we'll be doing something different with this in the final analysis, so can't be bothered to write code I'm fairly sure will be thrown away :)
